### PR TITLE
Add NAS KWS model, Dynamic Augmentation and Automated Evaluation Notebook

### PR DESCRIPTION
**Pull request by alicangok**
### Major changes:
- **Dynamic Augmentation** is introduced
   - Instead of fixing the augmented examples during dataset creation, the dataset loader now generates unique training examples during each epoch, significantly boosting robustness against noise and time shifts.
   - The more costly "speed augmentation" remains fixed, carried out once during dataset creation.
   - For stability of validation results across epochs, the validation examples (original + augmentations) are also fixed; they are constructed during initial dataset creation.
- **Changed the dataset filename** (dataset2.pt->dataset3.pt) to avoid potential mix-ups, as this PR introduces a major change
   - Added "shift_limits" property to each sample (for possible future feature compatibility, regarding voice activity detection)
   - The generated dataset contains the following:
      - The original training samples from Google Speech Commands, and 2 augmented versions of each sample with different speeds.
      - Additional training samples from Librispeech as additional examples for the "background" class.
      - The original validation samples from Google Speech Commands, and 2 augmented versions of each sample with different speeds, time shifts, and added white noise.
      - The original test samples from Google Speech Commands without any augmentation.
- **Dataset creation is significantly faster** (90 mins -> 4 mins), thanks to more efficient operations done in batches.
- The network found via "**Neural Architecture Search**" is introduced, which significantly improves accuracy than its predecessors (v2 & v3), having a higher parameter count, slightly increased #MACs, and latency (3.2ms -> 3.9ms).
- From: @EyubogluMerve: Added **automated evaluation notebook** for specified noise types and SNR levels..
   - Added a new dataset (signalmixer.py)
   - Modified msnoise.py to:
      - include "Tradeshow" as another type of noise
      - carry out proper train/test splits

---

### Summary of Improvements:
Along with the [previous PR](https://github.com/MaximIntegratedAI/ai8x-training/pull/258), we have **improved the KWS20 accuracy from ~86.5% to 92.5%** on the validation set which includes augmented samples, and from 87.6% to 93.7% on the clean test set.

The impact of each change on the KWS20 accuracy are as follows:
- pytsmod tempo augmentation -> torchaudio speed augmentation: +1%
- v3 -> v2 model: +1.5%
- v2 -> NAS model: +2.5%
- Dynamic noise & shift augmentation: +1%
- Total: +6% Absolute change in accuracy, from 86.5%->92.5%
    - 44% decrease in error rates, with even more significant reduction in false alarm rates.

### CODE REVIEW

**Comment by alicangok** [comment]
Changed the pull request to "draft" mode: Awaiting requested changes at: https://github.com/alicangok/ai8x-training/pull/1

**Review by ermanok — COMMENTED** [review]
Added minor comments.

File: notebooks/KWS_Noise_Evaluation.ipynb [hunk 1]
```diff
@@ -0,0 +1,905 @@
...
+      "Class UNKNOWN: 2824 elements\n"
+     ]
+    }
+   ],
+   "source": [
+    "sn = SimpleNamespace()\n",
+    "sn.truncate_testset = False\n",
+    "sn.act_mode_8bit = False\n",
+    "\n",
+    "data_path = '/data_ssd'\n",
```
Comment (aniktash) [inline notebooks/KWS_Noise_Evaluation.ipynb:206]: should it not be /data?

**Review by aniktash — COMMENTED** [review]
Some minor text updates suggested

File: notebooks/KWS_Noise_Evaluation.ipynb (hunk 1 above)
Comment (alicangok) [reply notebooks/KWS_Noise_Evaluation.ipynb:206]: We had been using /data_ssd on our local workstation. I've made the change to /data, thank you.

**Review by MaximGorkem — COMMENTED** [review]
Only few small comments, looks and trains nice.

File: datasets/kws20.py [hunk 2]
```diff
@@ -498,22 +515,66 @@ def quantize_audio(data, num_bits=8, compand=False, mu=255):
             q_data = np.clip(q_data, 0, max_val)
         return np.uint8(q_data)
 
-    def __gen_datasets(self, exp_len=16384, row_len=128, overlap_ratio=0):
-        print('Generating dataset from raw data samples for the first time. ')
-        print('This process will take significant time (~60 minutes)...')
+    def energy_detector(self, audio, fs):
```
[the remark this answers is not in the document]
Comment (alicangok) [reply datasets/kws20.py]: `energy_detector` -> `get_audio_endpoints`

**Review by rotx-maxim — APPROVED** [review]
